docs/rfcs/rfc-034-auth.md

```diff
@@ -0,0 +1,889 @@
+
+```gql
+type User {
+  id: String!
+  name: String!
+}
+```
+
+NOTE: This definition uses the filtering proposals from [#2730](https://github.com/neo4j/graphql/pulls/2730).
+
+The complete definition for this type will be:
+
+```gql
+input StringWhere {
+  OR: [StringWhere!]
+  AND: [StringWhere!]
+  NOT: StringWhere
+  equals: String
+  in: [String!]
+  matches: String
+  contains: String
+  startsWith: String
+  endsWith: String
+}
+
+input StringListWhere {
+  OR: [StringListWhere!]
+  AND: [StringListWhere!]
+  NOT: StringListWhere
+  all: [String!]
+  some: [String!]
+  single: String
+}
+
+input JWTPayloadWhere {
+  OR: [JWTPayloadWhere!]
+  AND: [JWTPayloadWhere!]
+  NOT: JWTPayloadWhere
+  sub: StringWhere
+  roles: StringListWhere
+}
+
+input UserWhere {
+  OR: [UserWhere!]
+  AND: [UserWhere!]
+  NOT: UserWhere
+  id: StringWhere
+  name: StringWhere
+}
+
+input UserAuthorizationWhere {
+  OR: [UserAuthorizationWhere!]
+  AND: [UserAuthorizationWhere!]
+  NOT: UserAuthorizationWhere
+  jwtPayload: JWTPayloadWhere
+  node: UserWhere
+}
+
+enum AuthorizationFilterOperation {
+  READ
+  UPDATE
+  DELETE
+  CREATE_RELATIONSHIP
+  DELETE_RELATIONSHIP
+}
+
+input UserAuthorizationFilterRule {
+  operations: [AuthorizationFilterOperation!]! = [READ, UPDATE, DELETE, CREATE_RELATIONSHIP, DELETE_RELATIONSHIP]
+  requireAuthentication: Boolean! = true
+  where: UserAuthorizationWhere!
+}
+
+enum AuthorizationPreValidateOperation {
+  READ
+  CREATE
+  UPDATE
+  DELETE
+  CREATE_RELATIONSHIP
+  DELETE_RELATIONSHIP
+}
+
+input UserAuthorizationPreValidateRule {
+  operations: [AuthorizationPreValidateOperation!]! = [READ, UPDATE, DELETE, CREATE_RELATIONSHIP, DELETE_RELATIONSHIP]
+  requireAuthentication: Boolean! = true
+  where: UserAuthorizationWhere!
+}
+
+enum AuthorizationPostValidateOperation {
+  CREATE
+  UPDATE
+  DELETE
+  CREATE_RELATIONSHIP
+  DELETE_RELATIONSHIP
+}
+
+input UserAuthorizationPostValidateRule {
+  operations: [AuthorizationPostValidateOperation!]! = [CREATE, UPDATE, DELETE, CREATE_RELATIONSHIP, DELETE_RELATIONSHIP]
+  requireAuthentication: Boolean! = true
+  where: UserAuthorizationWhere!
+}
+
+input UserAuthorizationValidateRules {
+  pre: [UserAuthorizationPreValidateRule!]
+  post: [UserAuthorizationPostValidateRule!]
+}
+
+directive @authorization(
+  filter: [UserAuthorizationFilterRule!]
+  validate: UserAuthorizationValidateRules
+) on OBJECT | FIELD_DEFINITION | INTERFACE
+```
+
+When an instance of the library is instantiated, for each type with an `@authorization` directive, the input types from above will be generated in isolation. The `@authorization` directive will then be validated against the generated input types. This will happen for each type with the directive.
+
+Points to note from above:
+
+* Each lifecycle hook has different operation compatibility:
+  * Filter:
+    * READ
+    * UPDATE
+    * DELETE
+    * CREATE_RELATIONSHIP
+    * DELETE_RELATIONSHIP
+  * Pre validate:
+    * READ
```

Comment:
Is it worth considering allowing `validate.pre`/`filter` rules to be applied for CREATE operations? If not it is impossible to prevent nested data from being accessed in CREATE mutations

Reply:
Even with this being the case, `CREATE` does not imply `READ` - if a user has a rule for `CREATE` _and_ `READ`, then we will apply the filter and pre-validate rules before returning data. If they have a rule only for `CREATE`, then what is documented here applies.